File: knowledge_graph/knowledge_graph/graph_utils.py
# Copyright 2023 Miguel Ángel González Santamarta

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
@file graph_utils.py
@brief Utility functions for Knowledge Graph operations.

This module provides helper functions for creating and manipulating
graph nodes, edges, content, and properties.
"""

from typing import Any, Callable, Dict, List, Optional, Union

from knowledge_graph_msgs.msg import Node, Edge, Content, Property


# Type mappings for content types
_TYPE_TO_STRING: Dict[int, str] = {
    Content.BOOL: "bool",
    Content.INT: "int",
    Content.FLOAT: "float",
    Content.DOUBLE: "double",
    Content.STRING: "string",
    Content.VBOOL: "bool[]",
    Content.VINT: "int[]",
    Content.VFLOAT: "float[]",
    Content.VDOUBLE: "double[]",
    Content.VSTRING: "string[]",
    Content.ERROR: "error",
}

_STRING_TO_TYPE: Dict[str, int] = {v: k for k, v in _TYPE_TO_STRING.items()}


def new_node(node_name: str, node_class: str) -> Node:
    """
    @brief Creates a new graph node.

    @param node_name The name of the node.
    @param node_class The class/type of the node.
    @return A new Node message with the specified name and class.
    """
    msg = Node()
    msg.node_name = node_name
    msg.node_class = node_class
    return msg


def new_edge(edge_class: str, edge_source: str, edge_target: str) -> Edge:
    """
    @brief Creates a new graph edge.

    @param edge_class The class/type of the edge.
    @param edge_source The name of the source node.
    @param edge_target The name of the target node.
    @return A new Edge message with the specified properties.
    """
    msg = Edge()
    msg.edge_class = edge_class
    msg.source_node = edge_source
    msg.target_node = edge_target
    return msg


def new_content(content: Any) -> Content:
    """
    @brief Creates a new Content message from a value.

    @param content The content value (bool, int, float, str, or list of these).
    @return A Content message with the appropriate type set.

    @note For lists, the type is inferred from the first element.
          Empty lists result in ERROR type.
    """
    msg = Content()

    if isinstance(content, bool):
        msg.bool_value = content
        msg.type = Content.BOOL
    elif isinstance(content, int):
        msg.int_value = content
        msg.type = Content.INT
    elif isinstance(content, float):
        msg.double_value = content
        msg.type = Content.DOUBLE
    elif isinstance(content, str):
        msg.string_value = content
        msg.type = Content.STRING
    elif isinstance(content, list) and len(content) > 0:
        first = content[0]
        if isinstance(first, bool):
            msg.bool_vector = content
            msg.type = Content.VBOOL
        elif isinstance(first, int):
            msg.int_vector = content
            msg.type = Content.VINT
        elif isinstance(first, float):
            msg.double_vector = content
            msg.type = Content.VDOUBLE
        elif isinstance(first, str):
            msg.string_vector = content
            msg.type = Content.VSTRING
        else:
            msg.type = Content.ERROR
    else:
        msg.type = Content.ERROR

    return msg


# Content type to getter mapping
_CONTENT_GETTERS: Dict[int, Callable[[Content], Any]] = {
    Content.BOOL: lambda c: c.bool_value,
    Content.INT: lambda c: c.int_value,
    Content.DOUBLE: lambda c: c.double_value,
    Content.STRING: lambda c: c.string_value,
    Content.VBOOL: lambda c: c.bool_vector,
    Content.VINT: lambda c: c.int_vector,
    Content.VDOUBLE: lambda c: c.double_vector,
    Content.VSTRING: lambda c: c.string_vector,
}


def get_content(content: Content) -> Optional[Any]:
    """
    @brief Extracts the value from a Content message.

    @param content The Content message to extract from.
    @return The extracted value, or None if type is ERROR or unknown.
    """
    getter = _CONTENT_GETTERS.get(content.type)
    return getter(content) if getter else None


def _find_property(element: Union[Node, Edge], key: str) -> Optional[Property]:
    """
    @brief Finds a property by key in a node or edge.

    @param element The node or edge to search.
    @param key The property key to find.
    @return The Property if found, None otherwise.
    """
    for prop in element.properties:
        if prop.key == key:
            return prop
    return None


def add_property(element: Union[Node, Edge], key: str, content: Any) -> bool:
    """
    @brief Adds or updates a property in a node or edge.

    @param element The node or edge to modify.
    @param key The property key.
    @param content The property value.
    @return True if successful, False if the content type is ERROR.
    """
    new_content_msg = new_content(content)

    if new_content_msg.type == Content.ERROR:
        return False

    existing = _find_property(element, key)
    if existing:
        existing.value = new_content_msg
    else:
        prop = Property()
        prop.key = key
        prop.value = new_content_msg
        element.properties.append(prop)

    return True


def get_property(element: Union[Node, Edge], key: str) -> Optional[Any]:
    """
    @brief Retrieves a property value from a node or edge.

    @param element The node or edge to search.
    @param key The property key to find.
    @return The property value if found, None otherwise.
    """
    prop = _find_property(element, key)
    return get_content(prop.value) if prop else None


def get_property_type(element: Union[Node, Edge], key: str) -> int:
    """
    @brief Gets the type of a property from a node or edge.

    @param element The node or edge to search.
    @param key The property key to find.
    @return The Content type of the property, or ERROR if not found.
    """
    prop = _find_property(element, key)
    return prop.value.type if prop else Content.ERROR


def _vector_to_string(vec: List[Any], converter: Callable[[Any], str]) -> str:
    """
    @brief Helper function to convert a vector to string representation.

    @param vec The vector to convert.
    @param converter Function to convert each element to string.
    @return String representation of the vector.
    """
    items = " ".join(f" {converter(v)}" for v in vec)
    return f"[{items}]"


def type_to_string(content_type: int) -> str:
    """
    @brief Converts a Content type to its string representation.

    @param content_type The Content type value.
    @return String representation of the type (e.g., "bool", "int", "string[]").
    """
    return _TYPE_TO_STRING.get(content_type, "Unknown")


def content_to_string(content: Content) -> str:
    """
    @brief Converts a Content message to its string representation.

    @param content The Content message to convert.
    @return String representation of the content value.
    """
    converters: Dict[int, Callable[[], str]] = {
        Content.BOOL: lambda: "true" if content.bool_value else "false",
        Content.INT: lambda: str(content.int_value),
        Content.FLOAT: lambda: str(content.float_value),
        Content.DOUBLE: lambda: str(content.double_value),
        Content.STRING: lambda: content.string_value,
        Content.VBOOL: lambda: _vector_to_string(
            content.bool_vector, lambda v: "true" if v else "false"
        ),
        Content.VINT: lambda: _vector_to_string(content.int_vector, str),
        Content.VFLOAT: lambda: _vector_to_string(content.float_vector, str),
        Content.VDOUBLE: lambda: _vector_to_string(content.double_vector, str),
        Content.VSTRING: lambda: _vector_to_string(content.string_vector, str),
    }

    converter = converters.get(content.type)
    return converter() if converter else "error"


def node_to_string(node: Node) -> str:
    """
    @brief Converts a Node message to its string representation.

    @param node The Node message to convert.
    @return String representation of the node with its properties.
    """
    lines = [f"{node.node_name} ({node.node_class})"]
    for prop in node.properties:
        lines.append(f"\t{prop.key}: [{content_to_string(prop.value)}]")
    return "\n".join(lines)


def edge_to_string(edge: Edge) -> str:
    """
    @brief Converts an Edge message to its string representation.

    @param edge The Edge message to convert.
    @return String representation of the edge.
    """
    return f" [{edge.edge_class}]{edge.source_node} -> {edge.target_node}"


def type_from_string(type_str: str) -> int:
    """
    @brief Converts a type string to its Content type value.

    @param type_str The string representation of the type.
    @return The Content type value, or ERROR if not recognized.
    """
    return _STRING_TO_TYPE.get(type_str, Content.ERROR)

File: knowledge_graph/src/knowledge_graph/graph_utils.cpp
// Copyright 2024 Universidad Politecnica de Madrid
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "knowledge_graph/graph_utils.hpp"

namespace knowledge_graph
{

knowledge_graph_msgs::msg::Node new_node(
  const std::string & node_name,
  const std::string & node_class)
{
  knowledge_graph_msgs::msg::Node ret;
  ret.node_name = node_name;
  ret.node_class = node_class;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content new_content<bool>(const bool & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.bool_value = content;
  ret.type = knowledge_graph_msgs::msg::Content::BOOL;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content new_content<int>(const int & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.int_value = content;
  ret.type = knowledge_graph_msgs::msg::Content::INT;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content new_content<float>(const float & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.float_value = content;
  ret.type = knowledge_graph_msgs::msg::Content::FLOAT;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content new_content<double>(const double & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.double_value = content;
  ret.type = knowledge_graph_msgs::msg::Content::DOUBLE;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content
new_content<std::string>(const std::string & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.string_value = content;
  ret.type = knowledge_graph_msgs::msg::Content::STRING;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content
new_content<std::vector<bool>>(const std::vector<bool> & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.bool_vector = content;
  ret.type = knowledge_graph_msgs::msg::Content::VBOOL;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content
new_content<std::vector<int>>(const std::vector<int> & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.int_vector = content;
  ret.type = knowledge_graph_msgs::msg::Content::VINT;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content
new_content<std::vector<float>>(const std::vector<float> & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.float_vector = content;
  ret.type = knowledge_graph_msgs::msg::Content::VFLOAT;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content
new_content<std::vector<double>>(const std::vector<double> & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.double_vector = content;
  ret.type = knowledge_graph_msgs::msg::Content::VDOUBLE;
  return ret;
}

template<>
knowledge_graph_msgs::msg::Content
new_content<std::vector<std::string>>(const std::vector<std::string> & content)
{
  knowledge_graph_msgs::msg::Content ret;
  ret.string_vector = content;
  ret.type = knowledge_graph_msgs::msg::Content::VSTRING;
  return ret;
}

knowledge_graph_msgs::msg::Edge new_edge(
  const std::string & edge_class,
  const std::string & edge_source,
  const std::string & edge_target)
{
  knowledge_graph_msgs::msg::Edge ret;
  ret.edge_class = edge_class;
  ret.source_node = edge_source;
  ret.target_node = edge_target;
  return ret;
}

template<>
std::optional<bool>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::BOOL) {
    return content.bool_value;
  } else {
    return {};
  }
}

template<>
std::optional<int>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::INT) {
    return content.int_value;
  } else {
    return {};
  }
}

template<>
std::optional<float>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::FLOAT) {
    return content.float_value;
  } else {
    return {};
  }
}

template<>
std::optional<double>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::DOUBLE) {
    return content.double_value;
  } else {
    return {};
  }
}

template<>
std::optional<std::string>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::STRING) {
    return content.string_value;
  } else {
    return {};
  }
}

template<>
std::optional<std::vector<bool>>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::VBOOL) {
    return content.bool_vector;
  } else {
    return {};
  }
}

template<>
std::optional<std::vector<int>>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::VINT) {
    return content.int_vector;
  } else {
    return {};
  }
}

template<>
std::optional<std::vector<float>>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::VFLOAT) {
    return content.float_vector;
  } else {
    return {};
  }
}

template<>
std::optional<std::vector<double>>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::VDOUBLE) {
    return content.double_vector;
  } else {
    return {};
  }
}

template<>
std::optional<std::vector<std::string>>
get_content(const knowledge_graph_msgs::msg::Content & content)
{
  if (content.type == knowledge_graph_msgs::msg::Content::VSTRING) {
    return content.string_vector;
  } else {
    return {};
  }
}

uint8_t
get_property_type(
  std::vector<knowledge_graph_msgs::msg::Property> & properties,
  const std::string key)
{
  auto it = properties.begin();
  while (it != properties.end()) {
    if (it->key == key) {
      return it->value.type;
    } else {
      ++it;
    }
  }
  return knowledge_graph_msgs::msg::Content::ERROR;
}

uint8_t get_property_type(
  knowledge_graph_msgs::msg::Node & node,
  const std::string key)
{
  return get_property_type(node.properties, key);
}

uint8_t get_property_type(
  knowledge_graph_msgs::msg::Edge & edge,
  const std::string key)
{
  return get_property_type(edge.properties, key);
}

std::string to_string(uint8_t edge_type)
{
  switch (edge_type) {
    case knowledge_graph_msgs::msg::Content::BOOL:
      return "bool";
      break;
    case knowledge_graph_msgs::msg::Content::INT:
      return "int";
      break;
    case knowledge_graph_msgs::msg::Content::FLOAT:
      return "float";
      break;
    case knowledge_graph_msgs::msg::Content::DOUBLE:
      return "double";
      break;
    case knowledge_graph_msgs::msg::Content::STRING:
      return "string";
      break;
    case knowledge_graph_msgs::msg::Content::VBOOL:
      return "bool[]";
      break;
    case knowledge_graph_msgs::msg::Content::VINT:
      return "int[]";
      break;
    case knowledge_graph_msgs::msg::Content::VFLOAT:
      return "float[]";
      break;
    case knowledge_graph_msgs::msg::Content::VDOUBLE:
      return "double[]";
      break;
    case knowledge_graph_msgs::msg::Content::VSTRING:
      return "string[]";
      break;
    case knowledge_graph_msgs::msg::Content::ERROR:
      return "error";
      break;
    default:
      return "Unknown";
      break;
  }
}

std::string to_string(const knowledge_graph_msgs::msg::Content & content)
{
  std::string ret;

  if (content.type == knowledge_graph_msgs::msg::Content::BOOL) {
    ret = content.bool_value ? "true" : "false";
  } else if (content.type == knowledge_graph_msgs::msg::Content::INT) {
    ret = std::to_string(content.int_value);
  } else if (content.type == knowledge_graph_msgs::msg::Content::FLOAT) {
    ret = std::to_string(content.float_value);
  } else if (content.type == knowledge_graph_msgs::msg::Content::DOUBLE) {
    ret = std::to_string(content.double_value);
  } else if (content.type == knowledge_graph_msgs::msg::Content::STRING) {
    ret = content.string_value;
  } else if (content.type == knowledge_graph_msgs::msg::Content::VBOOL) {
    ret = "[";
    for (const auto value : content.bool_vector) {
      ret = ret + " " + (value ? "true" : "false");
    }
    ret = ret + "]";
  } else if (content.type == knowledge_graph_msgs::msg::Content::VINT) {
    ret = "[";
    for (const auto value : content.int_vector) {
      ret = ret + " " + std::to_string(value);
    }
    ret = ret + "]";
  } else if (content.type == knowledge_graph_msgs::msg::Content::VFLOAT) {
    ret = "[";
    for (const auto value : content.float_vector) {
      ret = ret + " " + std::to_string(value);
    }
    ret = ret + "]";
  } else if (content.type == knowledge_graph_msgs::msg::Content::VDOUBLE) {
    ret = "[";
    for (const auto value : content.double_vector) {
      ret = ret + " " + std::to_string(value);
    }
    ret = ret + "]";
  } else if (content.type == knowledge_graph_msgs::msg::Content::VSTRING) {
    ret = "[";
    for (const auto value : content.string_vector) {
      ret = ret + " " + value;
    }
    ret = ret + "]";
  } else {
    ret = "error";
  }

  return ret;
}

uint8_t type_from_string(const std::string & type)
{
  for (uint8_t i = 0; i < knowledge_graph_msgs::msg::Content::NUM_TYPES; i++) {
    if (type == to_string(i)) {
      return i;
    }
  }
  return knowledge_graph_msgs::msg::Content::ERROR;
}

std::string to_string(const knowledge_graph_msgs::msg::Node & node)
{
  std::string ret;
  ret = ret + node.node_name + " (" + node.node_class + ")";

  for (const auto & prop : node.properties) {
    ret = ret + "\n\t" + prop.key + ": [" + to_string(prop.value) + "]";
  }

  return ret;
}

std::string to_string(const knowledge_graph_msgs::msg::Edge & edge)
{
  std::string ret;
  ret = ret + " [" + edge.edge_class + "]" + edge.source_node + " -> " +
    edge.target_node;
  return ret;
}

bool add_property(
  knowledge_graph_msgs::msg::Node & node, const knowledge_graph_msgs::msg::Property & property)
{
  bool found = false;
  auto newc = property.value;
  auto key = property.key;
  auto & properties = node.properties;

  if (newc.type == knowledge_graph_msgs::msg::Content::ERROR) {
    std::cerr << "Adding a property of type ERROR" << std::endl;
    return false;
  }

  auto it = properties.begin();
  while (!found && it != properties.end()) {
    if (it->key == key) {
      found = true;
      it->value = newc;
    } else {
      ++it;
    }
  }

  if (!found) {
    knowledge_graph_msgs::msg::Property prop;
    prop.key = key;
    prop.value = newc;
    properties.push_back(prop);
  }

  return true;
}


bool add_property(
  knowledge_graph_msgs::msg::Node & node,
  const std::vector<knowledge_graph_msgs::msg::Property> & property_vec)
{
  bool all_good = true;
  for (auto & prop : property_vec) {
    all_good |= add_property(node, prop);
  }
  return all_good;
}
bool add_property(
  knowledge_graph_msgs::msg::Edge & edge, const knowledge_graph_msgs::msg::Property & property)
{
  bool found = false;
  auto newc = property.value;
  auto key = property.key;
  auto & properties = edge.properties;

  if (newc.type == knowledge_graph_msgs::msg::Content::ERROR) {
    std::cerr << "Adding a property of type ERROR" << std::endl;
    return false;
  }

  auto it = properties.begin();
  while (!found && it != properties.end()) {
    if (it->key == key) {
      found = true;
      it->value = newc;
    } else {
      ++it;
    }
  }

  if (!found) {
    knowledge_graph_msgs::msg::Property prop;
    prop.key = key;
    prop.value = newc;
    properties.push_back(prop);
  }

  return true;
}


bool add_property(
  knowledge_graph_msgs::msg::Edge & edge,
  const std::vector<knowledge_graph_msgs::msg::Property> & property_vec)
{
  bool all_good = true;
  for (auto & prop : property_vec) {
    all_good |= add_property(edge, prop);
  }
  return all_good;
}

} // namespace knowledge_graph
